src/state_point.F90

```diff
@@ -126,151 +147,197 @@ subroutine write_state_point()
-                 group="tallies/tally" // trim(to_str(i)) // "/moments")
-            k = k + 1
-          case (SCORE_SCATTER_PN, SCORE_NU_SCATTER_PN)
-            do n_order = 0, t % moment_order(k)
-              moment_name = 'P' // trim(to_str(n_order))
+      if (n_tallies > 0) then
+
+        ! Print list of tally IDs
+        allocate(id_array(n_tallies))
+        allocate(key_array(n_tallies))
+
+        ! Write all tally information except results
+        do i = 1, n_tallies
+          tally => tallies(i)
+          key_array(i) = tally % id
+          id_array(i) = i
+        end do
+
+        call sp % write_data(id_array, "ids", &
+             group="tallies", length=n_tallies)
+        call sp % write_data(key_array, "keys", &
+             group="tallies", length=n_tallies)
+
+        deallocate(key_array)
+
+        ! Write all tally information except results
+        TALLY_METADATA: do i = 1, n_tallies
+
+          ! Get pointer to tally
+          tally => tallies(i)
+
+          call sp % write_data(len(tally % label), "label_size", &
+               group="tallies/tally " // trim(to_str(tally % id)))
+          if (len(tally % label) > 0) then
+            call sp % write_data(tally % label, "label", &
+                 group="tallies/tally " // trim(to_str(tally % id)))
+          endif
+
+          call sp % write_data(tally % estimator, "estimator", &
+               group="tallies/tally " // trim(to_str(tally % id)))
+          call sp % write_data(tally % n_realizations, "n_realizations", &
+               group="tallies/tally " // trim(to_str(tally % id)))
+          call sp % write_data(tally % n_filters, "n_filters", &
+               group="tallies/tally " // trim(to_str(tally % id)))
+
+          ! Write filter information
+          FILTER_LOOP: do j = 1, tally % n_filters
+
+            call sp % write_data(tally % filters(j) % type, "type", &
+                 group="tallies/tally " // trim(to_str(tally % id)) // &
+                 "/filter " // to_str(j))
+            call sp % write_data(tally % filters(j) % n_bins, "n_bins", &
+                 group="tallies/tally " // trim(to_str(tally % id)) // &
+                 "/filter " // to_str(j))
+            if (tally % filters(j) % type == FILTER_ENERGYIN .or. &
+                tally % filters(j) % type == FILTER_ENERGYOUT) then
+              call sp % write_data(tally % filters(j) % real_bins, "bins", &
+                   group="tallies/tally " // trim(to_str(tally % id)) // &
+                   "/filter " // to_str(j), &
+                   length=size(tally % filters(j) % real_bins))
+            else
+              call sp % write_data(tally % filters(j) % int_bins, "bins", &
+                   group="tallies/tally " // trim(to_str(tally % id)) // &
+                   "/filter " // to_str(j), &
+                   length=size(tally % filters(j) % int_bins))
+            end if
```

Comment:
I infer this to be a go-ahead for Sam's idea for to use tally mesh ids. However, I don't think it is appropriate to make a complete switch from array indices to user-supplied ids is in this PR. It is can be a pain to maintain a PR with the latest version of the "develop" branch over many months of review, particularly there are major differences in the ``state_point.F90`` file. I tried this before and on many occasions it allowed Git to incorrectly merge code in that file such that lots of the tests would break (*e.g.*, the strict ordering for binary statepoint files can easily be broken, which can take a non-trivial amount of time to debug). Hence, I reverted back to the array indices in the statepoint files, but with a ``StatePoint.link_with_summary(...)`` routine to utilize the user-supplied ids already reported in the HDF5 summary files.